# ANN model Input checker

use the ANN model optimiser code. then use maybe the most important features, and then second most important. and then run code to try different combinations. maybe get the MAE or some metric that will really help. Get the metric across the entier dataset

In [83]:
import pandas as pd
import numpy as np
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import tensorflow as tf
from keras.layers import Dense

columns_to_include = list(range(10))  # This creates a list of integers from 0 to 9
dataset = pd.read_csv('combined_data.csv', delimiter=';', usecols=columns_to_include)

dataset['Timestamp'] = pd.to_datetime(dataset['Timestamp'], format='%Y/%m/%d %H:%M')


In [84]:
X_temp = dataset.drop(columns=['Timestamp', dataset.columns[-1]])

y = dataset[dataset.columns[-1]]


In [85]:
# Use Random Forest to determine feature importance
rf = RandomForestRegressor()
rf.fit(X_temp, y)
feature_importances = rf.feature_importances_

most_important_feature_indices = np.argsort(feature_importances)[-7:]
most_important_feature_names = X_temp.columns[most_important_feature_indices]

print("Seven Most Important Features:", most_important_feature_names)

# Use only the 7 most important features for training and testing
X_most_important = X_temp.iloc[:, most_important_feature_indices]


Seven Most Important Features: Index(['Wind Speed', 'Wind Direction', 'Humidity', 'Temperature', 'Pressure',
       'Module Temperature', 'Irradiance'],
      dtype='object')


In [86]:
X_train, X_test, y_train, y_test = train_test_split(X_most_important, y, test_size=0.25, random_state=42)


In [87]:
# Standardize the features
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)


In [88]:
# Create and train the neural network
model = tf.keras.models.Sequential([
    Dense(64, activation='relu', input_dim=7),
    Dense(32, activation='relu'),
    Dense(1)
])

model.compile(loss='mse', optimizer='adam', metrics=['RootMeanSquaredError'])
model.summary()

Model: "sequential_10"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_31 (Dense)            (None, 64)                512       
                                                                 
 dense_32 (Dense)            (None, 32)                2080      
                                                                 
 dense_33 (Dense)            (None, 1)                 33        
                                                                 
Total params: 2625 (10.25 KB)
Trainable params: 2625 (10.25 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [89]:
# Train the model
model.fit(X_train, y_train, batch_size=32, epochs=500, verbose=2)


Epoch 1/500
204/204 - 4s - loss: 1679.4897 - root_mean_squared_error: 40.9816 - 4s/epoch - 20ms/step
Epoch 2/500
204/204 - 1s - loss: 354.2250 - root_mean_squared_error: 18.8209 - 609ms/epoch - 3ms/step
Epoch 3/500
204/204 - 1s - loss: 265.2425 - root_mean_squared_error: 16.2863 - 593ms/epoch - 3ms/step
Epoch 4/500
204/204 - 1s - loss: 232.2253 - root_mean_squared_error: 15.2389 - 562ms/epoch - 3ms/step
Epoch 5/500
204/204 - 1s - loss: 218.1814 - root_mean_squared_error: 14.7710 - 598ms/epoch - 3ms/step
Epoch 6/500
204/204 - 1s - loss: 211.1688 - root_mean_squared_error: 14.5316 - 619ms/epoch - 3ms/step
Epoch 7/500
204/204 - 1s - loss: 207.2176 - root_mean_squared_error: 14.3951 - 617ms/epoch - 3ms/step
Epoch 8/500
204/204 - 1s - loss: 204.6817 - root_mean_squared_error: 14.3067 - 577ms/epoch - 3ms/step
Epoch 9/500
204/204 - 1s - loss: 203.8835 - root_mean_squared_error: 14.2788 - 816ms/epoch - 4ms/step
Epoch 10/500
204/204 - 1s - loss: 202.1113 - root_mean_squared_error: 14.2166 - 875

Epoch 81/500
204/204 - 1s - loss: 177.0270 - root_mean_squared_error: 13.3052 - 877ms/epoch - 4ms/step
Epoch 82/500
204/204 - 1s - loss: 178.2389 - root_mean_squared_error: 13.3506 - 877ms/epoch - 4ms/step
Epoch 83/500
204/204 - 1s - loss: 176.8850 - root_mean_squared_error: 13.2998 - 853ms/epoch - 4ms/step
Epoch 84/500
204/204 - 1s - loss: 177.3673 - root_mean_squared_error: 13.3179 - 743ms/epoch - 4ms/step
Epoch 85/500
204/204 - 1s - loss: 177.0312 - root_mean_squared_error: 13.3053 - 880ms/epoch - 4ms/step
Epoch 86/500
204/204 - 1s - loss: 175.4655 - root_mean_squared_error: 13.2463 - 892ms/epoch - 4ms/step
Epoch 87/500
204/204 - 1s - loss: 175.9608 - root_mean_squared_error: 13.2650 - 959ms/epoch - 5ms/step
Epoch 88/500
204/204 - 1s - loss: 177.0655 - root_mean_squared_error: 13.3066 - 966ms/epoch - 5ms/step
Epoch 89/500
204/204 - 1s - loss: 175.3525 - root_mean_squared_error: 13.2421 - 965ms/epoch - 5ms/step
Epoch 90/500
204/204 - 1s - loss: 175.0770 - root_mean_squared_error: 13.

Epoch 160/500
204/204 - 1s - loss: 165.8196 - root_mean_squared_error: 12.8771 - 602ms/epoch - 3ms/step
Epoch 161/500
204/204 - 1s - loss: 165.1894 - root_mean_squared_error: 12.8526 - 598ms/epoch - 3ms/step
Epoch 162/500
204/204 - 1s - loss: 164.9675 - root_mean_squared_error: 12.8440 - 619ms/epoch - 3ms/step
Epoch 163/500
204/204 - 1s - loss: 164.6582 - root_mean_squared_error: 12.8319 - 599ms/epoch - 3ms/step
Epoch 164/500
204/204 - 1s - loss: 164.4258 - root_mean_squared_error: 12.8229 - 591ms/epoch - 3ms/step
Epoch 165/500
204/204 - 1s - loss: 163.7931 - root_mean_squared_error: 12.7982 - 764ms/epoch - 4ms/step
Epoch 166/500
204/204 - 1s - loss: 166.3664 - root_mean_squared_error: 12.8983 - 886ms/epoch - 4ms/step
Epoch 167/500
204/204 - 1s - loss: 164.8982 - root_mean_squared_error: 12.8413 - 883ms/epoch - 4ms/step
Epoch 168/500
204/204 - 1s - loss: 164.7768 - root_mean_squared_error: 12.8365 - 894ms/epoch - 4ms/step
Epoch 169/500
204/204 - 1s - loss: 164.8566 - root_mean_squared_

Epoch 239/500
204/204 - 1s - loss: 158.5540 - root_mean_squared_error: 12.5918 - 655ms/epoch - 3ms/step
Epoch 240/500
204/204 - 1s - loss: 157.8837 - root_mean_squared_error: 12.5652 - 626ms/epoch - 3ms/step
Epoch 241/500
204/204 - 1s - loss: 158.1510 - root_mean_squared_error: 12.5758 - 878ms/epoch - 4ms/step
Epoch 242/500
204/204 - 1s - loss: 158.4254 - root_mean_squared_error: 12.5867 - 874ms/epoch - 4ms/step
Epoch 243/500
204/204 - 1s - loss: 157.4780 - root_mean_squared_error: 12.5490 - 853ms/epoch - 4ms/step
Epoch 244/500
204/204 - 1s - loss: 157.7546 - root_mean_squared_error: 12.5600 - 929ms/epoch - 5ms/step
Epoch 245/500
204/204 - 1s - loss: 155.8876 - root_mean_squared_error: 12.4855 - 767ms/epoch - 4ms/step
Epoch 246/500
204/204 - 1s - loss: 156.8651 - root_mean_squared_error: 12.5246 - 601ms/epoch - 3ms/step
Epoch 247/500
204/204 - 1s - loss: 156.8181 - root_mean_squared_error: 12.5227 - 594ms/epoch - 3ms/step
Epoch 248/500
204/204 - 1s - loss: 157.6936 - root_mean_squared_

Epoch 318/500
204/204 - 1s - loss: 151.0242 - root_mean_squared_error: 12.2892 - 552ms/epoch - 3ms/step
Epoch 319/500
204/204 - 1s - loss: 150.1312 - root_mean_squared_error: 12.2528 - 526ms/epoch - 3ms/step
Epoch 320/500
204/204 - 1s - loss: 151.7357 - root_mean_squared_error: 12.3181 - 602ms/epoch - 3ms/step
Epoch 321/500
204/204 - 1s - loss: 151.5580 - root_mean_squared_error: 12.3109 - 592ms/epoch - 3ms/step
Epoch 322/500
204/204 - 1s - loss: 150.4019 - root_mean_squared_error: 12.2638 - 591ms/epoch - 3ms/step
Epoch 323/500
204/204 - 1s - loss: 151.5460 - root_mean_squared_error: 12.3104 - 551ms/epoch - 3ms/step
Epoch 324/500
204/204 - 1s - loss: 149.6561 - root_mean_squared_error: 12.2334 - 581ms/epoch - 3ms/step
Epoch 325/500
204/204 - 1s - loss: 149.8133 - root_mean_squared_error: 12.2398 - 604ms/epoch - 3ms/step
Epoch 326/500
204/204 - 1s - loss: 150.5715 - root_mean_squared_error: 12.2708 - 690ms/epoch - 3ms/step
Epoch 327/500
204/204 - 1s - loss: 150.2887 - root_mean_squared_

Epoch 397/500
204/204 - 1s - loss: 145.2023 - root_mean_squared_error: 12.0500 - 775ms/epoch - 4ms/step
Epoch 398/500
204/204 - 1s - loss: 144.9460 - root_mean_squared_error: 12.0394 - 590ms/epoch - 3ms/step
Epoch 399/500
204/204 - 1s - loss: 144.7365 - root_mean_squared_error: 12.0306 - 611ms/epoch - 3ms/step
Epoch 400/500
204/204 - 1s - loss: 146.0135 - root_mean_squared_error: 12.0836 - 601ms/epoch - 3ms/step
Epoch 401/500
204/204 - 1s - loss: 145.2948 - root_mean_squared_error: 12.0538 - 601ms/epoch - 3ms/step
Epoch 402/500
204/204 - 1s - loss: 144.2534 - root_mean_squared_error: 12.0106 - 597ms/epoch - 3ms/step
Epoch 403/500
204/204 - 1s - loss: 145.1561 - root_mean_squared_error: 12.0481 - 579ms/epoch - 3ms/step
Epoch 404/500
204/204 - 1s - loss: 145.2082 - root_mean_squared_error: 12.0502 - 584ms/epoch - 3ms/step
Epoch 405/500
204/204 - 1s - loss: 145.1590 - root_mean_squared_error: 12.0482 - 662ms/epoch - 3ms/step
Epoch 406/500
204/204 - 1s - loss: 144.8354 - root_mean_squared_

Epoch 476/500
204/204 - 1s - loss: 140.8849 - root_mean_squared_error: 11.8695 - 602ms/epoch - 3ms/step
Epoch 477/500
204/204 - 1s - loss: 140.6845 - root_mean_squared_error: 11.8610 - 581ms/epoch - 3ms/step
Epoch 478/500
204/204 - 1s - loss: 140.4545 - root_mean_squared_error: 11.8514 - 555ms/epoch - 3ms/step
Epoch 479/500
204/204 - 1s - loss: 139.5547 - root_mean_squared_error: 11.8133 - 606ms/epoch - 3ms/step
Epoch 480/500
204/204 - 1s - loss: 139.6522 - root_mean_squared_error: 11.8175 - 585ms/epoch - 3ms/step
Epoch 481/500
204/204 - 1s - loss: 139.8118 - root_mean_squared_error: 11.8242 - 594ms/epoch - 3ms/step
Epoch 482/500
204/204 - 1s - loss: 139.7182 - root_mean_squared_error: 11.8202 - 593ms/epoch - 3ms/step
Epoch 483/500
204/204 - 1s - loss: 140.7978 - root_mean_squared_error: 11.8658 - 592ms/epoch - 3ms/step
Epoch 484/500
204/204 - 1s - loss: 140.1713 - root_mean_squared_error: 11.8394 - 592ms/epoch - 3ms/step
Epoch 485/500
204/204 - 1s - loss: 140.5394 - root_mean_squared_

In [90]:
# Evaluate the model
test_loss, test_rmse = model.evaluate(X_test, y_test)
print("Root Mean Squared Error on Test Data:", test_rmse)

68/68 [==============================] - 1s 6ms/step - loss: 204.5220 - root_mean_squared_error: 14.3011
Root Mean Squared Error on Test Data: 14.301119804382324


Results: 

All features: 
loss: 205.6688 - root_mean_squared_error: 14.3412
Root Mean Squared Error on Test Data: 14.341156959533691


Irradiance only: 
loss: 230.1525 - root_mean_squared_error: 15.1708
Root Mean Squared Error on Test Data: 15.1707763671875

Irradiance and Module temp:
loss: 231.9447 - root_mean_squared_error: 15.2297
Root Mean Squared Error on Test Data: 15.229729652404785

Irradiance, Module temp and pressure:
loss: 232.4498 - root_mean_squared_error: 15.2463
Root Mean Squared Error on Test Data: 15.24630355834961

Irradiance, Module temp, pressure and Temperature:
loss: 219.7063 - root_mean_squared_error: 14.8225
Root Mean Squared Error on Test Data: 14.822494506835938

Irradiance, Module temp, pressure, Temperature and Humidity:
loss: 222.6722 - root_mean_squared_error: 14.9222
Root Mean Squared Error on Test Data: 14.92220401763916

Irradiance, Module temp, pressure, Temperature, Humidity and Wind Direction:
loss: 211.1915 - root_mean_squared_error: 14.5324
Root Mean Squared Error on Test Data: 14.532430648803711


Irradiance, Module temp, pressure, Temperature, Humidity, wind Speed and Wind Direction:
loss: 204.5220 - root_mean_squared_error: 14.3011
Root Mean Squared Error on Test Data: 14.301119804382324
